packages/waku/src/router/client.ts

```diff
@@ -102,19 +103,46 @@ type ChangeRoute = (
   },
 ) => Promise<void>;
 
+type ChangeRouteEvent = 'start' | 'complete';
+
+type ChangeRouteCallback = (
+  route: RouteProps,
+  options: {
+    shouldScroll: boolean;
+    skipRefetch?: boolean;
+  },
+) => void;
+
+const EMPTY_ROUTE_CHANGE_LISTENERS: Record<
+  ChangeRouteEvent,
+  Set<ChangeRouteCallback>
+> = { start: new Set(), complete: new Set() };
+
 type PrefetchRoute = (route: RouteProps) => void;
 
 const RouterContext = createContext<{
   route: RouteProps;
   changeRoute: ChangeRoute;
   prefetchRoute: PrefetchRoute;
+  routeChangeListeners: Record<ChangeRouteEvent, Set<ChangeRouteCallback>>;
 } | null>(null);
 
 export function useRouter() {
   const router = useContext(RouterContext);
   if (!router) {
     throw new Error('Missing Router');
   }
+
+  const unstable_events = useMemo(() => {
```

Comment:
We are using `useRef` internally anyway, so technically, we can avoid useMemo here and useCallback for executeListeners.
I'm not sure if the final result is more readable or less.
I would do something like this:
```diff
- const routeChangeListenersRef = useRef(EMPTY_ROUTE_CHANGE_LISTENERS);
+ const routeChangeListenersRef = useRef(null);
+ if (routeChangeListenersRef.current === null) {
+   const listeners = { start: new Set(), complete: new Set() };
+   const events = ...;
+   const executeLisetners = ...;
+   routeChangeListenersRef.current = [events, executeListenrs];
+ }
```

Reply:
changed